accepted/0000-sbom-command.md

```diff
@@ -0,0 +1,566 @@
+
+Initially, we'll support the most widely used versions of the SPDX and CycloneDX specifications (likely v2.3 for SPDX and v1.5 for CycloneDX). Best effort will be made to support new versions as they gain adoption across the ecosystem.
+
+
+## Rationale and Alternatives
+
+There are a few existing tools which can be used to generate an SBOM from an npm project:
+
+* <code>[@cyclonedex/cyclonedx-npm](https://www.npmjs.com/package/@cyclonedx/cyclonedx-npm)</code> - A CLI for generating a CycloneDX-style SBOM from an npm project. This project is written in TypeScript and actually invokes <code>npm-ls</code> in order to get dependency information for the project.
+* <code>[spdx-sbom-generator](https://github.com/opensbom-generator/spdx-sbom-generator)</code> - A CLI for generating SPDX-style SBOMs for a number of different package managers (including npm). Currently, this tool only works with npm projects using <em>lockfileVersion</em> 1 so it’s not viable for a large number of projects (current <em>lockfileVersion</em> is 3)
+
+While you can effectively generate the same output we’re proposing with this combination of tools, there is value in having this capability supported directly in npm. Beyond the obvious developer-experience benefit of having SBOM generation baked-in to the CLI, it gives us a future path to do things like automatic-signing of SBOMs or integration of SBOMs into the package publishing process.
+
+
+## Implementation
+
+The `npm-sbom` command will use <code>[arborist](https://github.com/npm/cli/tree/latest/workspaces/arborist)</code> to construct the dependency tree for the current project and then invoke `querySelectorAll` to select the set of nodes to be included in the SBOM.
+
+### Errors
+
+When using the `node_modules` to render the SBOM (i.e. when NOT using the `--package-lock-only` flag) any of the following conditions will cause an error to be reported and prevent the SBOM from being generated:
+
+- Any missing dependencies which are NOT marked as optional
+- Any invalid dependencies (e.g. a mismatch between the `package-lock.json` and the `node_modules`)
+
+### Format Details
+
+Both of the SBOM formats present a flat list of dependencies (CycloneDX groups these under a key named `components` while SPDX groups them under a key named `packages`). The following sections describe how a dependency will be presented for the different SBOM formats.
+
+
+#### CycloneDX
+
+```json
+{
+  "type": "library",
+  "name": "debug",
+  "version": "4.3.4",
+  "bom-ref": "debug@4.3.4",
+  "purl": "pkg:npm/debug@4.3.4",
+  "scope": "required",
+  "externalReferences": [
+    {
+      "type": "distribution",
+      "url": "https://registry.npmjs.org/debug/-/debug-4.3.4.tgz"
+    }
+  ],
+  "properties": [
+    {
+      "name": "cdx:npm:package:path",
+      "value": "node_modules/debug"
+    }
+  ],
+  "hashes": [
```

Comment:
these hashes probably belong to the `externalReferences` of type `distribution`,
not to the component.

the component is a folder in `node_modules` dir , ...